lib/pacemaker/pcmk_sched_group.c

```diff
@@ -661,16 +665,26 @@ pcmk__group_apply_location(pcmk_resource_t *rsc, pcmk__location_t *location)
 
         pcmk_resource_t *member = (pcmk_resource_t *) iter->data;
 
-        member->private->cmds->apply_location(member, location);
-
-        if (reset_scores) {
-            /* The first member of colocated groups needs to use the original
-             * node scores, but subsequent members should work on a copy, since
-             * the first member's scores already incorporate theirs.
+        if (pe__group_flag_is_set(rsc, pcmk__group_colocated)
+            && (iter != rsc->private->children)) {
+            /* The first member gets the constraint score added to its affected
+             * allowed nodes. However we use a 0 score for subsequent members,
+             * because otherwise when assigning the member, the constraint score
+             * would be counted multiple times (once for each later member) due
+             * to internal group colocations. Though the 0 score will not affect
+             * later members' allowed node scores, it ensures that affected
+             * nodes are in each member's allowed nodes, enabling the member on
+             * those nodes in asymmetric clusters.
+             *
+             * This is suboptimal if a group member other than the first is
+             * explicitly colocated with another resource. In that case, when
+             * the colocation primary is assigned, it will not consider the
+             * group's location preferences.
              */
-            reset_scores = false;
             location->nodes = node_list_copy;
         }
+
+        member->private->cmds->apply_location(member, location);
```

Comment:
The effect is the same as before: previously, it would call apply_location() for the first member using the actual scores, then (in the same loop iteration) change the node list to the zero-score copy, and that would remain in effect for all later iterations.

Now,

   rsc != rsc->private->parent->private->children->data

ensures the zero-score copy is used for members after the first. We're assigning nodes at the start of the second iteration (and later iterations, though that's redundant) rather than the end of the first iteration.

Reply:
Ah you're right, I misread. I think the redundant assignments in later iterations are what threw me off.